src/Components/Selection.re

```diff
@@ -0,0 +1,53 @@
+[@deriving show({with_path: false})]
+type t = {
+  anchor: int,
+  focus: int,
+};
+
+let initial: t = {anchor: 0, focus: 0};
+
+let anchor = (selection: t): int => {
+  selection.anchor;
+};
+
+let focus = (selection: t): int => {
+  selection.focus;
+};
+
+let withinLength = (text: string, position: int): int => {
+  let withinStart = max(position, 0);
+  let withinBoth = min(withinStart, String.length(text));
+
+  withinBoth;
+};
+
+let create = (~text: string, ~anchor: int, ~focus: int): t => {
+  let safeAnchor = withinLength(text, anchor);
+  let safeFocus = withinLength(text, focus);
+
+  {anchor: safeAnchor, focus: safeFocus};
+};
+
+let rangeWidth = (selection: t): int => {
+  abs(selection.focus - selection.anchor);
+};
+
+let rangeStart = (selection: t): int => {
+  min(selection.focus, selection.anchor);
+};
+
+let rangeEnd = (selection: t): int => {
+  max(selection.focus, selection.anchor);
+};
```

Comment:
Suggestion for better names, since I still don't understand what "range" means in this context. I use `Left` and `Right` instead of `Start` and `End` because the latter is at best inaccurate. The "start" of the selection is the anchor, and the "end" is the node.

```suggestion
let length = (selection: t): int => {
  abs(selection.focus - selection.anchor);
};

let offsetLeft = (selection: t): int => {
  min(selection.focus, selection.anchor);
};

let offsetRight = (selection: t): int => {
  max(selection.focus, selection.anchor);
};
```